ocaml/typing/typedecl.ml

```diff
@@ -362,11 +362,14 @@ let transl_labels ~new_var_jkind env univars closed lbls =
     Builtin_attributes.warning_scope attrs
       (fun () ->
          let gbl =
-           match mut with
-           | Mutable -> Mode.Global_flag.Global
-           | Immutable -> Typemode.transl_global_flags
```

Comment:
I think you mean that some consumers of the `ld_modalities` field would want to check if it contains `Global`. But I think that's bad behaviour anyway: modalities is just a sequence of functions that should be applied; analysis based on the presence/absence of certain modalities is bound to be unreliable. Similarly, `ld_mutable` is another mapping of modes.

The current approach is to bake the behaviour of `Global` into `Mutable`, that doesn't sound fundamentally wrong.

Reply:
`ld_mutable` isn't a mapping it is a constant mode, and only some consumers of `ld_modalities` should also consume `ld_mutable` -- e.g. field projection doesn't need to read it. It is important to put the right information in the right places.

Currently, each consumer of `ld_modalities` is having to do an additional check of `ld_mutable` in order to find out the actual modality of the field. That is needlessly error prone. It also means that the semantics of `ld_modalities` is unclear: it's some subset of the modalities on the field that varies based on the syntax used by the user.

As an example of a misbehaviour caused by this confusion, I think that the following is currently incorrectly rejected with this PR:
```ocaml
module M : sig
  type t = { mutable foo : string }
end = struct
  type t = { mutable global foo : string }
end
```
Obviously we are going to change the meaning of `mutable` later so that this will fail, but I think it is accepted currently and this PR accidentally breaks it.

Reply:
Yes good point that `ld_mutable` is better understood as a constant that can be easily compared with each other. To preserve its simplicity, we should avoid adding modality logic to it. I will fix that.